myhoard/web_server.py

```diff
@@ -78,6 +80,42 @@ async def backup_list(self, _request):
                     response["backups"] = self.controller.state["backups"]
                 return json_response(response)
 
+    async def backup_preserve(self, request):
+        with self._handle_request(name="backup_preserve"):
+            body = await self._get_request_json(request)
+            stream_id = body.get("stream_id")
+            preserve_until = body.get("preserve_until")
+
+            if stream_id is None or preserve_until is None:
+                raise BadRequest("`stream_id` and `preserve_until` must be set in request body.")
+
+            try:
+                preserve_until = datetime.fromisoformat(preserve_until).astimezone(timezone.utc)
+                now = datetime.now(timezone.utc)
+                if preserve_until < now:
+                    raise BadRequest("`preserve_until` must be a date in the future.")
+            except ValueError:
+                raise BadRequest("`preserve_until` must be a valid isoformat datetime string.")
+
+            return self._handle_preservation_request(PreservationRequest(stream_id=stream_id, preserve_until=preserve_until))
+
+    async def backup_cancel_preserve(self, request):
+        with self._handle_request(name="backup_cancel_preserve"):
+            body = await self._get_request_json(request)
+            stream_id = body.get("stream_id")
+            if stream_id is None:
+                raise BadRequest("`stream_id` must be set in request body.")
+
+            return self._handle_preservation_request(PreservationRequest(stream_id=stream_id, preserve_until=None))
+
+    def _handle_preservation_request(self, preservation_request: PreservationRequest):
+        try:
+            with self.controller.lock:
+                self.controller.mark_backup_preservation(preservation_request)
```

Comment:
This blocks the asyncio thread and makes the API non-functional. We should not be doing any blocking operations here. Also holding the controller lock throughout the entire operation is bad, the lock is intended to only be held while checking what current state is and updating the state, not while performing network IO. Presumably this should just update the controller state to indicate preservation request has been issued and then wait for the operation to completely like we do elsewhere.

Reply:
:thinking: Though shouldn't the controller be locked in order to avoid backup deletion while updating the preservation file? Also, requesting preservation should refresh the backup list.

Reply:
Tho I am noticing we are actually not consistent with this :/ Seems we acquire the controller lock a couple of times in the API (e.g `backup_list`)  and we are not acquiring the lock when calling `_refresh_backups_list`... it updates the controller state without acquiring the lock. Unless I am understanding wrong your comment :)

Reply:
For backup deletion `_purge_old_backups` should just check the pending preservation requests and not delete a backup that has been preserved.

Looks like `_refresh_backup_list()` is missing lock around the `self.state_manager.update_state(backups=backups, backups_fetched_at=time.time())` call but I don't think it needs anything beyond that.

Reply:
yes, though `_purge_old_backups` is removing the backup stream without acquiring the lock (it later acquires it for updating owned stream ids). So we might reach a race condition when trying to preserve it... stream is still valid but at the moment of preservation it suddenly got removed, which might lead to a false positive.

I'll open a separate PR for `_refresh_backup_list` lock acquisition, that should remove the lock acquisition in `GET /backup`

Reply:
> though _purge_old_backups is removing the backup stream without acquiring the lock

We can make that update some local state to indicate that we're already in the process of removing a backup and if a preservation request is done for a backup we're already deleting, the request fails.